### PR TITLE
boot: Stop linking against libefi.a

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

File: src/boot/efi/disk.c [hunk 1]
```diff
@@ -32,7 +32,27 @@ EFI_STATUS disk_get_part_uuid(EFI_HANDLE *handle, char16_t uuid[static 37]) {
                 if (hd.SignatureType != SIGNATURE_TYPE_GUID)
                         continue;
 
-                GuidToString(uuid, (EFI_GUID *) &hd.Signature);
+                _cleanup_free_ char16_t *tmp = xasprintf(
```
Comment (daandemeyer) [inline src/boot/efi/disk.c:35]: OOM check?

File: src/boot/efi/util.c [hunk 2]
```diff
@@ -44,15 +45,10 @@ EFI_STATUS efivar_set(const EFI_GUID *vendor, const char16_t *name, const char16
...
-        char16_t str[32];
-
         assert(vendor);
         assert(name);
 
-        /* Note that SPrint has no native sized length specifier and will always use ValueToString()
-         * regardless of what sign we tell it to use. Therefore, UINTN_MAX will come out as -1 on
-         * 64bit machines. */
-        ValueToString(str, false, i);
+        _cleanup_free_ char16_t *str = xasprintf("%zu", i);
```
Comment (daandemeyer) [inline src/boot/efi/util.c:51]: OOM check?

File: src/boot/efi/util.h [hunk 3]
```diff
@@ -116,14 +116,23 @@ static inline void unload_imagep(EFI_HANDLE *image) {
                 (void) BS->UnloadImage(*image);
 }
 
+/* Creates a EFI_GUID pointer suitable for EFI APIs. Use of const allows the compiler to merge multiple
+ * uses (although, currently compilers do that regardless). Most EFI APIs declare their EFI_GUID input
+ * as non-const, but almost all of them are in fact const. */
+#define MAKE_GUID_PTR(name) ((EFI_GUID *) &(const EFI_GUID) name##_GUID)
```
Comment (daandemeyer) [inline src/boot/efi/util.h:122]: I'd shorten this to `GUID_PTR()`

File: src/boot/efi/boot.c [hunk 4]
```diff
@@ -2041,8 +2043,8 @@ static EFI_STATUS boot_windows_bitlocker(void) {
 
                 if (streq16((char16_t *) (buf + offset), L"Windows Boot Manager")) {
                         err = efivar_set_raw(
-                                EFI_GLOBAL_GUID,
-                                L"BootNext",
+                                MAKE_GUID_PTR(EFI_GLOBAL_VARIABLE),
+                                u"BootNext",
```
Comment (daandemeyer) [inline src/boot/efi/boot.c:2047]: This seems unrelated to this commit. Should probably be somewhere else or its own commit

File: src/boot/efi/util.c [hunk 5]
```diff
@@ -515,10 +515,14 @@ uint64_t get_os_indications_supported(void) {
 }
 
 #ifdef EFI_DEBUG
-__attribute__((noinline)) void debug_break(void) {
+extern uint8_t _text, _data;
+__attribute__((noinline)) void notify_debugger(const char *identity, volatile bool wait) {
+        printf("%s@%p,%p\n", identity, &_text, &_data);
+        if(wait)
```
Comment (bluca) [inline src/boot/efi/util.c]: missing space

File: src/boot/efi/secure-boot.c [hunk 6]
```diff
@@ -10,7 +10,7 @@ bool secure_boot_enabled(void) {
         bool secure = false;  /* avoid false maybe-uninitialized warning */
         EFI_STATUS err;
 
-        err = efivar_get_boolean_u8(EFI_GLOBAL_GUID, L"SecureBoot", &secure);
+        err = efivar_get_boolean_u8(MAKE_GUID_PTR(EFI_GLOBAL_VARIABLE), u"SecureBoot", &secure);
```
Comment (daandemeyer) [inline src/boot/efi/secure-boot.c:13]: Ditto, unrelated change, here and below

File: src/boot/efi/stub.c [hunk 7]
```diff
@@ -417,12 +417,4 @@ static EFI_STATUS real_main(EFI_HANDLE image) {
...
-EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
-        InitializeLib(image, sys_table);
-
-        notify_debugger("systemd-stub", /*wait_for_debugger=*/false);
-
-        EFI_STATUS err = real_main(image);
-        log_wait();
-        return err;
-}
+DEFINE_EFI_MAIN_FUNCTION(real_main, "systemd-stub", /*wait_for_debugger=*/false);
```
Comment (daandemeyer) [inline src/boot/efi/stub.c]: Can we rename `real_main()` to `run()` here and elsewhere? That's consistent with our other tooling

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks good, just a few things

File: src/boot/efi/disk.c (hunk 1 above)
Comment (medhefgo) [reply src/boot/efi/disk.c:35]: Our efi code panics on allocation failures, hence the functions that start with `x`.

File: src/boot/efi/boot.c (hunk 4 above)
Comment (medhefgo) [reply src/boot/efi/boot.c:2047]: It's porting to the new way of using GUIDs, so I think it's relevant. `EFI_GLOBAL_GUID` uses the old style and a non-standard (compared to the spec) name for this var. Might as well do both at once…
Whether these are all touched in one commit or split out makes little difference imho. They're gonna be commit noise anyways.

File: src/boot/efi/util.h (hunk 3 above)
Comment (medhefgo) [reply src/boot/efi/util.h:122]: I certainly wouldn't mind too much, but do we do similar stuff in userspace that produces compound literals? The "make" conveys that something is created/"allocated" while `GUID_PTR` sounds like a cast to me…

More opinions on this would be nice.

**Comment by bluca** [comment]
@medhefgo could this be causing this build failure on armhf (arm32)? Just seen it after uploading 253-rc2 to debian experimental:

```
/usr/bin/cc -o src/boot/efi/linuxarm.elf.stub -DGNU_EFI_USE_MS_ABI -DSD_BOOT=1 -ffreestanding -fshort-wchar -fvisibility=hidden -I '/<<PKGBUILDDIR>>/src/fundamental' -I '/<<PKGBUILDDIR>>/src/boot/efi' -include src/boot/efi/efi_config.h -include version.h -isystem /usr/include/efi/arm -isystem /usr/include/efi -std=gnu11 -Wall -Wextra -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=format-signedness -Werror=implicit-function-declaration -Werror=implicit-int -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wno-maybe-uninitialized -Wno-unused-result -ftrivial-auto-var-init=zero -fno-stack-protector -fno-strict-aliasing -fpic -fwide-exec-charset=UCS2 -mgeneral-regs-only -flto=auto -O2 '-ffile-prefix-map=/<<PKGBUILDDIR>>=.' -fuse-ld=bfd -L /usr/lib -nostdlib -T /usr/lib/elf_arm_efi.lds -Wl,--build-id=sha1 -Wl,--fatal-warnings -Wl,--no-undefined -Wl,--warn-common -Wl,-Bsymbolic -z nocombreloc /usr/lib/crt0-efi-arm.o -Wl,--no-warn-execstack -Wl,--no-warn-rwx-segments -shared -Wl,--defsym=EFI_SUBSYSTEM=0xa -Wl,--no-wchar-size-warning src/boot/efi/bootspec-fundamental.c.o src/boot/efi/efivars-fundamental.c.o src/boot/efi/sha256.c.o src/boot/efi/string-util-fundamental.c.o src/boot/efi/tpm-pcr.c.o src/boot/efi/console.c.o src/boot/efi/devicetree.c.o src/boot/efi/drivers.c.o src/boot/efi/disk.c.o src/boot/efi/efi-string.c.o src/boot/efi/graphics.c.o src/boot/efi/initrd.c.o src/boot/efi/log.c.o src/boot/efi/measure.c.o src/boot/efi/part-discovery.c.o src/boot/efi/pe.c.o src/boot/efi/random-seed.c.o src/boot/efi/secure-boot.c.o src/boot/efi/ticks.c.o src/boot/efi/util.c.o src/boot/efi/vmm.c.o src/boot/efi/cpio.c.o src/boot/efi/linux.c.o src/boot/efi/splash.c.o src/boot/efi/stub.c.o -lgnuefi -lgcc
/usr/bin/ld.bfd: /usr/lib/gcc/arm-linux-gnueabihf/12/libgcc.a(_dvmd_lnx.o): in function `__aeabi_ldiv0':
(.text+0x6): undefined reference to `raise'
```

**Comment by medhefgo** [comment]
Yes it does. libgcc's implementation for `__aeabi_ldiv0` intrinsic will call `raise` (as defined in userspace) if a division by zero happens. We can just provide a definition of `raise` ourselves for the moment.

**Comment by bluca** [comment]
> Yes it does. libgcc's implementation for `__aeabi_ldiv0` intrinsic will call `raise` (as defined in userspace) if a division by zero happens. We can just provide a definition of `raise` ourselves for the moment.

I see - are you able to provide a fix? arm32 is supported in debian/ubuntu, so it would be a regression to ship 253 with this issue, although I'm not exactly sure how many arm32 users of sd-stub/sd-boot are out there
